core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -371,13 +371,10 @@ private[chisel3] object Builder {
     */
   def pushPrefix(d: HasId): Boolean = {
     def buildAggName(id: HasId): Option[String] = {
-      // TODO This is slow, can we store this information upon binding?
-      def getSubName(field: Data, parent: Data): Option[String] = parent match {
-        case vec: Vec[_] =>
-          val idx = vec.indexOf(field)
-          if (idx >= 0) Some(idx.toString) else None
-        case rec: Record => rec.elements.collectFirst { case (name, d) if d == field => name }
-        case _ => Builder.exception(s"Shouldn't see non-Aggregate $parent as parent in ChildBinding!")
+      def getSubName(field: Data): Option[String] = field.getOptionRef.flatMap {
+        case Slot(_, field) => Some(field) // Record
+        case Index(_, ILit(n)) => Some(n.toString) // Vec
+        case _ => None // Vec dynamic indexing
```

Comment:
Can this check be made stricter and fail noisily on unexpected cases?

Reply:
This is a better naming path where failure leads to it just not supplying a prefix for names. Do you really think it should fail noisily here?

Reply:
I don't fully understand the ref system and lifecycle, but the overall principle I'm getting at is that we should have code bugs fail as noisily as possible. If the no-name path is expected and intended, then this is fine - though can it be made tighter than `_`?

Reply:
I'll add a tighter bound